addons/hr_holidays/models/hr_leave_accrual_plan_level.py

```diff
@@ -306,11 +327,29 @@ def _get_previous_date(self, last_call):
             else:
                 return last_call + relativedelta(years=-1, month=second_month, day=self.second_month_day)
         elif self.frequency == 'yearly':
-            month = MONTHS.index(self.yearly_month) + 1
-            year_date = last_call + relativedelta(month=month, day=self.yearly_day)
-            if last_call >= year_date:
-                return year_date
+            if self.accrual_date_ref == "reference":
+                month = MONTHS.index(self.yearly_month) + 1
+                year_date = last_call + relativedelta(month=month, day=self.yearly_day)
+                if last_call >= year_date:
+                    return year_date
+                else:
+                    return last_call + relativedelta(years=-1, month=month, day=self.yearly_day)
             else:
-                return last_call + relativedelta(years=-1, month=month, day=self.yearly_day)
+                month, day = self._get_accrual_month_day_from_allocation_date(allocation_start_date)
+                year_date = last_call + relativedelta(month=month, day=day)
+                if last_call >= year_date:
+                    return year_date
+                else:
+                    return last_call + relativedelta(years=-1, month=month, day=day)
         else:
             return False
+
+    def _get_accrual_month_day_from_allocation_date(self, allocation_start_date):
```

Comment:
This doesn't work in batch
```suggestion
    def _get_accrual_month_day_from_allocation_date(self, allocation_start_date):
        self.ensure_one()
```

Reply:
Ah indeed, if its already called by the other methods that should be fine without it :)
To know when to use `ensure_one`, or when to make a method work in batch, there's sadly no universal solution to that; I'd say that as much as possible, we should try to make methods work in batch, especially those making queries (`search`, `self.env.ref`, `read_group`,...) in order to reduce the amount of total queries and make everything work faster.
Sometimes, like in this case, the  methods aren't doing queries, and are a bit clearer without the addition of a loop and giving the results inside a dictionary, in that case we use ensure_one allowing us to address self :)
Hope this helps